packages/cli/test/utils/simulation/assertions.ts

```diff
@@ -0,0 +1,307 @@
+          );
+          expect(currentJustifiedSlot + 1).to.equal(
+            finalizedSlot,
+            `there should be no gaps between finalized and previous justified epochs, received finalized ${finalizedSlot} and previous ${previousJustifiedSlot}`
+          );
+        });
+      });
+    }
+  });
+}
+
+export function forkAssertions(env: SimulationEnvironment): void {
+  describe("forks", () => {
+    for (let n = 0; n < env.params.beaconNodes; n++) {
+      describe(`beacon node "${n}"`, () => {
+        describe("altair fork", () => {
+          it("should occur on right slot", async () => {
+            // const node = env.nodes[n];
+
+            if (env.params.altairEpoch === Infinity) {
+              return;
+            }
+
+            const expectedSlot = env.params.altairEpoch * env.params.slotsPerEpoch + env.params.genesisSlotsDelay;
+            await env.clock.waitForEndOfSlot(expectedSlot);
+
+            // await node.api.getBlock();
+            // Match if the block is from the right fork
+          });
+        });
+
+        describe("bellatrix fork", () => {
+          it("should occur on right slot", async () => {
+            // const node = env.nodes[n];
+
+            if (env.params.bellatrixEpoch === Infinity) {
+              return;
+            }
+
+            const expectedSlot = env.params.bellatrixEpoch * env.params.slotsPerEpoch + env.params.genesisSlotsDelay;
+            await env.clock.waitForEndOfSlot(expectedSlot);
+
+            // await node.api.getBlock();
+            // Match ic the block is from the right fork
+          });
+        });
+      });
+    }
+  });
+}
+
+export function nodeAssertions(env: SimulationEnvironment): void {
+  describe("node", () => {
+    it(`should have correct "${env.params.beaconNodes}" number of nodes`, () => {
+      expect(env.nodes.length).to.equal(env.params.beaconNodes);
+    });
+
+    for (let n = 0; n < env.params.beaconNodes; n++) {
+      describe(`beacon node "${n}"`, () => {
+        it("should have correct health status", async () => {
+          const node = env.nodes[n];
+          expect(await node.api.node.getHealth()).to.equal(200);
+        });
+
+        it("should be completely synced", async () => {
+          const node = env.nodes[n];
+          const syncStatus = await node.api.node.getSyncingStatus();
+
+          expect(syncStatus.data.isSyncing).to.equal(false);
+          expect(syncStatus.data.syncDistance).to.equal("0");
+        });
+
+        // NOT POSSIBLE
+        it.skip("should have correct number of validator clients");
+
+        describe("validator - n", () => {
+          it("should have correct status");
+          it("should have correct number of keys loaded");
+        });
+      });
+    }
+
+    it("all nodes should have same head", async () => {
+      const heads = await Promise.all(env.nodes.map((n) => n.api.beacon.getBlockHeader("head")));
```

Comment:
Nodes may receive blocks with some delay, so the head may be different without anything going wrong

Reply:
Yes, but in our constraint simulation environment we expect that nodes should have same heads. 

Is there any alternative approach you suggest to check this? Should we check for a slot rather the head?

Reply:
The slot is by definition the same, since all nodes have the same clock (they are in the same host). Here it's very important the timing of when you are querying this. It can happen that (say Slot N starts at 17:00:00)
- Node0: set head at time 17:00:00.012
- Node1: set head at time 17:00:00.803
- Node2: set head at time 17:00:00.445
- Node3: set head at time 17:00:00.116

If you call heads at 17:00:00 exactly, the result you are gonna get is uncertain. The specific point through a slot at which you call heads is crucial to prevent false positives